### PR TITLE
Change PickFirstLeafLoadBalancer to only have 1 subchannel at a time

### PR DESCRIPTION
**Pull request by larry-safran**
Hide behind environment variable GRPC_SERIALIZE_RETRIES

This is for testing with GMS core to see if they use the new PF with only 1 subchannel at a time trying to reconnect whether that eliminates their 6% data increase.

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 1]
```diff
@@ -64,17 +64,27 @@ final class PickFirstLeafLoadBalancer extends LoadBalancer {
...
-  private ScheduledHandle scheduleConnectionTask;
+  private ScheduledHandle scheduleConnectionTask = null;
   private ConnectivityState rawConnectivityState = IDLE;
   private ConnectivityState concludedState = IDLE;
-  private final boolean enableHappyEyeballs =
-      PickFirstLoadBalancerProvider.isEnabledHappyEyeballs();
+  private final boolean enableHappyEyeballs = !isSerializingRetries()
+      && PickFirstLoadBalancerProvider.isEnabledHappyEyeballs();
   private boolean notAPetiolePolicy = true; // means not under a petiole policy
+  private Status lastError = null;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:73]: Nothing reads this field.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 2]
```diff
@@ -304,6 +329,42 @@ void processSubchannelState(SubchannelData subchannelData, ConnectivityStateInfo
...
+      }
+    }
+
+    // Just allow the previous one to trigger when ready if we're already in backoff
+    if (reconnectTask != null) {
+      return;
+    }
+
+    if (reconnectPolicy == null) {
+      reconnectPolicy = bkoffPolProvider.get();
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:354]: I see the field is initialized here, but it isn't ever cleared on success. InternalSubchannel does that in cancelReconnectTask().

**Review by ejona86 — COMMENTED** [review]
Sending what I have. I'll need to look at this more later.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 3]
```diff
@@ -458,7 +538,8 @@ private SubchannelData createNewSubchannel(SocketAddress addr, Attributes attrs)
   }
 
   private boolean isPassComplete() {
-    if (addressIndex.isValid() || subchannels.size() < addressIndex.size()) {
+    if ((!serializingRetries && addressIndex.isValid())
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:541]: Why do we want to ignore whether the index is valid?

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 4]
```diff
@@ -391,8 +454,21 @@ public void requestConnection() {
...
+        }
         addressIndex.increment();
-        requestConnection();
+        if (!serializingRetries) {
+          requestConnection();
+        } else {
+          if (!addressIndex.isValid()) {
+            scheduleBackoff();
+          } else {
+            subchannelData.subchannel.shutdown(); // shutdown the previous subchannel
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:467]: Don't we need to shutdown the subchannel as soon as it goes TF? Otherwise while we try other subchannels this one will start reconnecting when its backoff expires.

Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:468]: Good point

Comment (ejona86) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:468]: I think the shutdown needs to be in `processSubchannelState()`, as that method could have already called `addressIndex.increment()`. But we also have the problem that when you shutdown the subchannel you need to remove it from `subchannels` (missing in the recent addition), which will then need other tweaks.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 3 above)
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:537]: Because we reset it for subchannel retry.

**Comment by larry-safran** [comment]
Ready for re-review

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 5]
```diff
@@ -370,7 +432,12 @@ private void shutdownRemaining(SubchannelData activeSubchannelData) {
    */
   @Override
   public void requestConnection() {
-    if (!addressIndex.isValid() || rawConnectivityState == SHUTDOWN) {
+    if (rawConnectivityState == SHUTDOWN) {
+      return;
+    }
+
+    if (!addressIndex.isValid()) {
+      scheduleBackoff();
```
[the remark this answers is not in the document]
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:437]: Seems this came back. The problem I have with this line is if it is ever needed (it actually schedules the timer), then that means there is a bug somewhere else. Because it means we should have been in backoff but didn't set the timer.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 6]
```diff
@@ -391,8 +455,19 @@ public void requestConnection() {
         scheduleNextConnection();
         break;
       case TRANSIENT_FAILURE:
+        if (scheduleConnectionTask != null) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:458]: Why was this added? scheduleConnectionTask is guaranteed to be null for serializing retry, so this seems an unrelated bug or unnecessary.

**Review by ejona86 — COMMENTED** [review]
Sending what I have. This seems much less bug-prone

**Review by ejona86 — COMMENTED** [review]
After the first failure for an address, I don't see how we start a second connection to the same address (after the backoff).

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 7]
```diff
@@ -77,6 +78,7 @@ final class InternalSubchannel implements InternalInstrumented<ChannelStats>, Tr
   private final CallTracer callsTracer;
   private final ChannelTracer channelTracer;
   private final ChannelLogger channelLogger;
+  private final boolean recconectDisabled;
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java:81]: Done

[the remark this answers is not in the document]
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: You got the "only two c's" right but not the "two n's" :smile:

File: xds/src/test/java/io/grpc/xds/RingHashLoadBalancerTest.java [hunk 8]
```diff
@@ -150,7 +150,8 @@ public void subchannelLazyConnectUntilPicked() {
     assertThat(result.getStatus().isOk()).isTrue();
     assertThat(result.getSubchannel()).isNull();
     Subchannel subchannel = Iterables.getOnlyElement(subchannels.values());
-    int expectedTimes = PickFirstLoadBalancerProvider.isEnabledHappyEyeballs() ? 1 : 2;
+    int expectedTimes = PickFirstLoadBalancerProvider.isEnabledHappyEyeballs()
+                            || !PickFirstLoadBalancerProvider.isEnabledNewPickFirst() ? 2 : 1;
```
Comment (ejona86) [inline xds/src/test/java/io/grpc/xds/RingHashLoadBalancerTest.java:154]: It doesn't seem the `isEnabledHappyEyeballs()` is really doing anything. Should this just be (aka, adjusting the original code to be s/isEnabledHappyEyeballs/isEnabledNewPickFirst/):
```java
PickFirstLoadBalancerProvider.isEnabledNewPickFirst() ? 1 : 2
```

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 3 above)
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:532]: This is still nagging at me. I feel like either we never need this condition or we need to replace it with something else for serializingRetries. I agree that we need this function to return true even though we've reset the index (for `++numTf`).

**Review by ejona86 — COMMENTED** [review]
Looks good other than the nagging `!serializingRetries && addressIndex.isValid()`

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 9]
```diff
@@ -277,10 +287,12 @@ void processSubchannelState(SubchannelData subchannelData, ConnectivityStateInfo
           if (addressIndex.increment()) {
             cancelScheduleTask();
             requestConnection(); // is recursive so might hit the end of the addresses
+          } else {
+            scheduleBackoff();
           }
         }
 
-        if (isPassComplete()) {
+        if (!firstPass || isPassComplete()) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:295]: Let's at least add a comment that the `!firstPass` is an optimization.

**Review by ejona86 — APPROVED** [review]

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 9 above)
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java]: Just removed it.  Looping on a boolean comparison hopefully won't take too long.
